components/match2/wikis/leagueoflegends/match_group_input_custom.lua

```diff
@@ -155,425 +144,286 @@ end
+			match = MatchGroupInput.readPlayersOfTeam(match, opponentIndex, opponent.name, {
+				resolveRedirect = true,
+				applyUnderScores = true,
+				maxNumPlayers = MAX_NUM_PLAYERS,
+			})
+		elseif opponent.type == Opponent.solo then
+			opponent.match2players = Json.parseIfString(opponent.match2players) or {}
+			opponent.match2players[1].name = opponent.name
 		end
+
+		return opponent
+	end)
+end
+
+---@param scoreInput string|number|nil
+---@return integer? #SCORE
+---@return string? #STATUS
+function MatchFunctions._parseScoreInput(scoreInput)
+	if not scoreInput then
+		return
 	end
 
-	--apply walkover input
-	match.walkover = string.upper(match.walkover or '')
-	if Logic.isNumeric(match.walkover) then
-		local winnerIndex = tonumber(match.walkover)
-		opponents = MatchFunctions._makeAllOpponentsLoseByWalkover(opponents, STATUS_DEFAULT_LOSS)
-		opponents[winnerIndex].status = STATUS_DEFAULT_WIN
-		match.finished = true
-	elseif Logic.isNumeric(match.winner) and Table.includes(ALLOWED_STATUSES, match.walkover) then
-		local winnerIndex = tonumber(match.winner)
-		opponents = MatchFunctions._makeAllOpponentsLoseByWalkover(opponents, match.walkover)
-		opponents[winnerIndex].status = STATUS_DEFAULT_WIN
-		match.finished = true
+	if Logic.isNumeric(scoreInput) then
+		return tonumber(scoreInput), MatchGroupInput.STATUS.SCORE
 	end
 
-	-- see if match should actually be finished if score is set
-	if not Logic.readBool(match.finished) then
-		MatchFunctions._finishMatch(match, opponents, isScoreSet)
+	local scoreUpperCase = string.upper(scoreInput)
+	if Table.includes(MatchGroupInput.STATUS_INPUTS, scoreUpperCase) then
+		return MatchGroupInput.SCORE_NOT_PLAYED, scoreUpperCase
 	end
+end
 
-	-- apply placements and winner if finshed
-	if Logic.readBool(match.finished) and not match.standaloneMatch then
-		match, opponents = CustomMatchGroupInput.getResultTypeAndWinner(match, opponents)
+---@param walkoverInput string #wikicode input
+---@param isWinner boolean
+---@return integer? #SCORE
+---@return string? #STATUS
+function MatchFunctions._opponentWalkover(walkoverInput, isWinner)
+	if Logic.isNumeric(walkoverInput) then
+		walkoverInput = MatchGroupInput.STATUS.DEFAULT_LOSS
 	end
 
-	-- Update all opponents with new values
-	for opponentIndex, opponent in pairs(opponents) do
-		match['opponent' .. opponentIndex] = opponent
+	local walkoverUpperCase = string.upper(walkoverInput)
+	if Table.includes(MatchGroupInput.STATUS_INPUTS, walkoverUpperCase) then
+		return MatchGroupInput.SCORE_NOT_PLAYED, isWinner and MatchGroupInput.STATUS.DEFAULT_WIN or walkoverUpperCase
 	end
-	return match
 end
 
 ---@param match table
----@param opponents table[]
----@param isScoreSet boolean
----@return table
-function MatchFunctions._finishMatch(match, opponents, isScoreSet)
+---@param opponents {score: integer?}[]
+---@return boolean
+function MatchFunctions._isFinished(match, opponents)
```

Comment:
maybe move to commons, making the check for special status an optional pass (use commons one if not passed)
probably need to be able to pass the thresholds too (if not passed use a default one from commons)

Reply:
I believe all wikis have the same thresholds, so was already thinking about having them force standardized

Reply:
if it is the same for all that makes it easier for sure :)

Reply:
Why do you think special status should be optional?

Reply:
not optional but overwritable if we have wikis handeling it differently
(currently the placementCheckSpecialStatus function is in the lol custom)

i was thinking something like this
```lua
---@param match table
---@param opponents {score: integer?}[]
---@param checkSpecialStatus? fun(opponents: {status: string?}[]): boolean
---@return boolean
function MatchGroupInput.isFinished(match, opponents, checkSpecialStatus)
	local finished = Logic.readBoolOrNil(match.finished)
	if finished ~= nil then
		return finished
	end

	-- If a winner has been set
	if Logic.isNotEmpty(match.winner) then
		return true
	end

	-- If special status has been applied to an opponent
	checkSpecialStatus = checkSpecialStatus or MatchGroupInput.checkSpecialStatus
	if CustomMatchGroupInput.placementCheckSpecialStatus(opponents) then
		return true
	end

	if not MatchGroupInput.hasScore(opponents) then
		return false
	end

	-- Check if all/enough games have been played. If they have, mark as finished
	local firstTo = math.floor(match.bestof / 2)
	if Array.any(opponents, function(opponent) return (tonumber(opponent.score) or 0) > firstTo end) then
		return true
	end
	local scoreSum = Array.reduce(opponents, function(sum, opponent) return sum + (opponent.score or 0) end, 0)
	if scoreSum >= match.bestof then
		return true
	end

	-- If enough time has passed since match started, it should be marked as finished
	local threshold = match.dateexact and SECONDS_UNTIL_FINISHED_EXACT
		or SECONDS_UNTIL_FINISHED_NOT_EXACT
	if match.timestamp ~= DateExt.defaultTimestamp and match.timestamp + threshold < NOW then
		return true
	end

	return false
end
```

Reply:
Removed the custom one on League. It was there because the map level used to indirectly call this function, but map level on LOL never set status field